### ISSUE TITLE
Feature Request?  multi-tenancy, with separate hieradata requires workaround.

### ISSUE THREAD

**Issue by brettswift** [puppetlabs/r10k#195]
I'm experimenting with multi-tenancy, and my use case is a bit unique.   This is a different approach to #101 if multiple applications (tenants) have the git-flow  release model for puppet code.

Goal: separate release cycles of applications along the git-flow branches. 

```
app1 control repo:  develop --> test --> staging --> prod
app1 hieradata repo:  develop --> test --> staging --> prod
app2 control repo:  develop --> test --> staging --> prod
app2 hieradata repo:  develop --> test --> staging --> prod
```

One way of doing this is locking each to a git hash (#101), but that can be cumbersome and requires everyone be on the same puppet file.

What I'd like to get working is multi-tenancy with r10k, so that the above workflow could work.   If hieradata is inside the same repo as the Puppetfile / aka _control repo_ then this is fine,  **but where it collides is separate hieradata repos**.
### Premise
- every "tenant" would have both a _control repo_ and a _hiera data_ repo.  
- **Hieradata** repo is also released in the git-flow style.  (We did this to allow easier hotfixing of data vs code). 
#### Single Tenant /etc/r10k.yaml

``` yaml
:sources:
  data:
    basedir: /etc/puppetlabs/puppet/hieradata
    prefix: false
    remote: ssh://git@stash/codo/hieradata.git
  puppet:
    basedir: /etc/puppetlabs/puppet/environments
    prefix: false
    remote: ssh://git@stash/codo/puppetconfig.git

:purgedirs:
  - /etc/puppetlabs/puppet/environments
  - /etc/puppetlabs/puppet/hieradata
```
#### multitenant /etc/r10k.yaml:

``` yaml
:sources:
  app1_h: 
    basedir: /etc/puppetlabs/puppet/hieradata
    prefix: true
    remote: ssh://git@stash/codo/hieradata.git
  app1_p:
    basedir: /etc/puppetlabs/puppet/environments
    prefix: true
    remote: ssh://git@stash/codo/puppetconfig.git

:purgedirs:
  - /etc/puppetlabs/puppet/environments
  - /etc/puppetlabs/puppet/hieradata
```

config that drives /etc/hiera.yaml

```
puppet::hiera_data_dir: "/etc/puppetlabs/puppet/hieradata/%{::environment}"
```

I need to set the prefix because that's how multi-tenancy differentiates tenants in the module path and hiera data path.  But now I can't just set my environment to "app1_develop" because on the master the environments are "app1_h_develop" and "app1_p_develop" (for hiera data, and 'puppet' control repo respectively). 

It's not valid yaml to get rid of the _p,  and _h.   

A rather dirty solution is to have: 
1) environment = `develop`
2) fact / enc set to `app1`
3) hiera.yaml data dir set to `$fact _h $environment`
4 modulepath set to `$fact _p $environment`

... but that's quite kludgy. 

@adrienthebo  does this make sense or am I missing something?  If it makes sense would a feature to implement **prefix_name** solve this? As follows: 

``` yaml
:sources:
  app1_hieradata: 
    basedir: /etc/puppetlabs/puppet/hieradata
    prefix: true
    prefix_name: app1
    remote: ssh://git@stash/codo/hieradata.git
  app1_puppetconfig:
    basedir: /etc/puppetlabs/puppet/environments
    prefix: true
    prefix_name: app1
    remote: ssh://git@stash/codo/puppetconfig.git

:purgedirs:
  - /etc/puppetlabs/puppet/environments
  - /etc/puppetlabs/puppet/hieradata
```

If i set /etc/r10k.yaml to have two sources with the same name, it won't create both as it obviously has duplicate keys.  

Do you have a better idea to solve this?

**Comment by brettswift** [comment]
I've put some code in for git, which I'll test monday.  If you want to see what I was talking about, here's the branch: 
https://github.com/brettswift/r10k/compare/feature/multi_tenant_prefix_name?expand=1

Thanks!

**Comment by adrienthebo** [comment]
In the very early days of r10k I prototyped out ideas where the prefix name could be interpolated based on some supplied scope, and I think this progresses in the same vein. So what sort of other branch naming schemes are there? There's `#{sourcename}_#{branchname}` and `#{prefixname}_#{branchname}`, are there any other reasonable branch naming conventions beyond these?

**Comment by brettswift** [comment]
I'm not really sure of other naming conventions.   I think `#{sourcename}_#{branchname}` solves things if hieradata is _inside_ your control repo, which would equal `#{branchname}` if prefix=> false. (same use case as I see it?)

`#{prefixname}_#{branchname}` would support environments when hieradata is separate. 

I come from only those two perspectives.  Would some PSE's have experience with another use case?

For this feature to be accepted, is it sufficient to only support this on .git (no svn)?  Any other code review comments to turn this into a pull request?

**Comment by adrienthebo** [comment]
Inside of the Git environment class there's the BranchName construct; I think it might suffice to extract that into a general class and reuse that from both SVN and Git. This would render the concern about SVN support vs Git support moot.

With respect to option naming, I kinda like the idea of allowing prefix to be `true` where the prefix is the source name, `String` where string is a specific prefix, and `false` where no prefixing will be done. It seems a bit cleaner rather than adding another option. What do you think?

**Comment by brettswift** [comment]
Hey, sorry it's taken me a bit to get back to you.

I agree with all of your suggestions.   However,  multiple tenancy with hieradata will have an issue due to this:  https://tickets.puppetlabs.com/browse/HI-46

So, I guess putting time into fixing this is kind of moot for my use case until that is fixed. 

We'd ultimately need the environment parameter to drive the hiera.yaml file so that multi-tenancy actually works.

**Comment by fiddyspence** [comment]
I have exactly the same issue here, but I solved it outside hiera as part of the deployment by just deep merging each discrete hiera tree into a single set of files from each set of modules.  We have an element of priority lookup - viz. later bits of the module path are overwritten by the earlier (higher priority) parts.  This works in our environment, but may not be a general use case (because danger will robinson)

**Comment by ghoneycutt** [comment]
@brettswift Though I've been pushing for real environment support for Hiera, I don't think that r10k has to wait on that. If you don't care about different backends per environment, you can setup your Hiera config to look in say `%{environment}/roles/%{role}` instead of `roles/%{role}`. r10k deploys to say /var/lib/hiera/${environment}. Though Puppet does not really support multi-tenancy yet, r10k can start. When puppet has better support for it, the change would likely just be changing the directories in which you deploy Hiera data.

**Comment by brettswift** [comment]
This is one of those solutions that is so simple is kind of embarrassing
lol.

I'm going to try this asap. And get this feature built.

Essentially move $::environment from hiera.yaml/datadir into each level of
the hierarchy. Simple!

It moves from ~250 yaml files in memory to 250*# of environments, currently
that means 3000+ on our puppet master. (there's a $hostname tier with a lot
of data). That shouldn't be much of an impact.

Much appreciated.

**Comment by adrienthebo** [comment]
See also: https://tickets.puppetlabs.com/browse/RK-35

**Comment by adrienthebo** [comment]
Merged in 910709a.
